### PR TITLE
[mlir] Rectify mishandling in `InsertOpConstantFolder` causing crash with assertion when using `mlir-opt --canonicalize`

### LINKED ISSUE #74236: [mlir] --canonicalize crashed with assertion failure "expected integer attribute type to equal element type".
git version: e9c6f3f5e7e23b23de4eeaa182ebfcb7d2188495

system: ```Ubuntu 20.04.6 LTS (Focal Fossa)```

reproduced with: ```mlir-opt --canonicalize a.mlir```

a.mlir:
```
llvm.func @malloc(i64) -> !llvm.ptr
func.func @func2(%arg0: index, %arg1: memref<13x13xi64>, %arg2: index) {
  %cst_7 = arith.constant dense<1526248407> : vector<1xi64>
  %1 = llvm.mlir.constant(1 : index) : i64
  %101 = vector.insert %1, %cst_7 [0] : i64 into vector<1xi64>
  vector.print %101 : vector<1xi64>
  return
}
```

trace:
```
mlir-opt: /data/bin/llvm-project/mlir/lib/IR/BuiltinAttributes.cpp:961: static mlir::DenseElementsAttr mlir::DenseElementsAttr::get(mlir::ShapedType, llvm::ArrayRef<mlir::Attribute>): Assertion `intAttr.getType() == eltType && "expected integer attribute type to equal element type"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /data/bin/llvm-project/build/bin/mlir-opt --inline a.mlir
 #0 0x0000560af72fee8f llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/data/bin/llvm-project/build/bin/mlir-opt+0x1640e8f)
 #1 0x0000560af72fbee4 SignalHandler(int) Signals.cpp:0:0
 #2 0x00007fbb1c867420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
 #3 0x00007fbb1c35200b raise (/lib/x86_64-linux-gnu/libc.so.6+0x4300b)
 #4 0x00007fbb1c331859 abort (/lib/x86_64-linux-gnu/libc.so.6+0x22859)
 #5 0x00007fbb1c331729 (/lib/x86_64-linux-gnu/libc.so.6+0x22729)
 #6 0x00007fbb1c342fd6 (/lib/x86_64-linux-gnu/libc.so.6+0x33fd6)
 #7 0x0000560afa0dc621 mlir::DenseElementsAttr::get(mlir::ShapedType, llvm::ArrayRef<mlir::Attribute>) (/data/bin/llvm-project/build/bin/mlir-opt+0x441e621)
 #8 0x0000560af93c28b6 (anonymous namespace)::InsertOpConstantFolder::matchAndRewrite(mlir::vector::InsertOp, mlir::PatternRewriter&) const VectorOps.cpp:0:0
 #9 0x0000560af762a7e8 mlir::detail::OpOrInterfaceRewritePatternBase<mlir::vector::InsertOp>::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&) const (/data/bin/llvm-project/build/bin/mlir-opt+0x196c7e8)
#10 0x0000560afccb3038 mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<mlir::LogicalResult (mlir::Pattern const&)>) (/data/bin/llvm-project/build/bin/mlir-opt+0x6ff5038)
#11 0x0000560afa0281cc (anonymous namespace)::GreedyPatternRewriteDriver::processWorklist() GreedyPatternRewriteDriver.cpp:0:0
#12 0x0000560afa02c63b mlir::applyPatternsAndFoldGreedily(mlir::Region&, mlir::FrozenRewritePatternSet const&, mlir::GreedyRewriteConfig, bool*) (/data/bin/llvm-project/build/bin/mlir-opt+0x436e63b)
#13 0x0000560af9f9190a (anonymous namespace)::Canonicalizer::runOnOperation() Canonicalizer.cpp:0:0
#14 0x0000560af9f7285e mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) (/data/bin/llvm-project/build/bin/mlir-opt+0x42b485e)
#15 0x0000560af9f72d3a mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (/data/bin/llvm-project/build/bin/mlir-opt+0x42b4d3a)
#16 0x0000560af9f73e03 mlir::LogicalResult llvm::function_ref<mlir::LogicalResult (mlir::OpPassManager&, mlir::Operation*)>::callback_fn<mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int)::'lambda'(mlir::OpPassManager&, mlir::Operation*)>(long, mlir::OpPassManager&, mlir::Operation*) Pass.cpp:0:0
#17 0x0000560af9fa1af3 (anonymous namespace)::InlinerPass::optimizeCallable(mlir::CallGraphNode*, llvm::StringMap<mlir::OpPassManager, llvm::MallocAllocator>&) Inliner.cpp:0:0
#18 0x0000560af9faaa26 mlir::LogicalResult llvm::function_ref<mlir::LogicalResult ((anonymous namespace)::CallGraphSCC&)>::callback_fn<(anonymous namespace)::InlinerPass::runOnOperation()::'lambda'((anonymous namespace)::CallGraphSCC&)>(long, (anonymous namespace)::CallGraphSCC&) Inliner.cpp:0:0
#19 0x0000560af9fab813 (anonymous namespace)::InlinerPass::runOnOperation() Inliner.cpp:0:0
#20 0x0000560af9f7285e mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) (/data/bin/llvm-project/build/bin/mlir-opt+0x42b485e)
#21 0x0000560af9f72d3a mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (/data/bin/llvm-project/build/bin/mlir-opt+0x42b4d3a)
#22 0x0000560af9f73564 mlir::PassManager::run(mlir::Operation*) (/data/bin/llvm-project/build/bin/mlir-opt+0x42b5564)
#23 0x0000560af9f642d3 performActions(llvm::raw_ostream&, std::shared_ptr<llvm::SourceMgr> const&, mlir::MLIRContext*, mlir::MlirOptMainConfig const&) MlirOptMain.cpp:0:0
#24 0x0000560af9f657a4 processBuffer(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::MlirOptMainConfig const&, mlir::DialectRegistry&, llvm::ThreadPool*) MlirOptMain.cpp:0:0
#25 0x0000560af9f65954 mlir::LogicalResult llvm::function_ref<mlir::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>::callback_fn<mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&)::'lambda'(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>(long, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&) MlirOptMain.cpp:0:0
#26 0x0000560afa064c44 mlir::splitAndProcessBuffer(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::function_ref<mlir::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>, llvm::raw_ostream&, bool, bool) (/data/bin/llvm-project/build/bin/mlir-opt+0x43a6c44)
#27 0x0000560af9f5e1b7 mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&) (/data/bin/llvm-project/build/bin/mlir-opt+0x42a01b7)
#28 0x0000560af9f65aa1 mlir::MlirOptMain(int, char**, llvm::StringRef, llvm::StringRef, mlir::DialectRegistry&) (/data/bin/llvm-project/build/bin/mlir-opt+0x42a7aa1)
#29 0x0000560af9f65f06 mlir::MlirOptMain(int, char**, llvm::StringRef, mlir::DialectRegistry&) (/data/bin/llvm-project/build/bin/mlir-opt+0x42a7f06)
#30 0x0000560af723650b main (/data/bin/llvm-project/build/bin/mlir-opt+0x157850b)
#31 0x00007fbb1c333083 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24083)
#32 0x0000560af72cf1ee _start (/data/bin/llvm-project/build/bin/mlir-opt+0x16111ee)
Aborted (core dumped)
```

### PR DESCRIPTION
**Pull request by Rajveer100**
Resolves #74236

### CODE REVIEW

**Comment by Rajveer100** [comment]
@joker-eph 
This currently breaks few tests since I have just put a `continue`, I think we probably want to check if there are two arguments only then we do so.

Let me know the direction for this.

**Comment by Rajveer100** [comment]
cc @kazutakahirata

**Review by stellaraccident — COMMENTED** [review]
Thank you for the patch and repro. There is something subtle going on here that I can't just spot in a code review. If others can't either, it will need someone familiar to step through it and fix properly vs just giving advice. In that case, probably needs an issue vs PR so we don't lose it.

**Comment by joker-eph** [comment]
It's not clear to me that the bug is in `DenseElementsAttr::get` actually, I would push the fix on the API call site instead.

However digging a bit more, the underlying problem seems that our folding hook tolerates that an attribute can be returned for an SSA Value with a type mismatch. This happens only for `llvm.constant in practice, because it accepts index types: `%1 = llvm.mlir.constant(1 : index) : i64`.
If we continue to accept this, then every folder should check that the attribute they get from matching a constant actually correspond to the expected SSA Type (in this case the fix would be in `InsertOpConstantFolder` for example).

**Comment by Rajveer100** [comment]
@stellaraccident @joker-eph

File: mlir/test/mlir-opt/issue-74236.mlir [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// RUN: mlir-opt -split-input-file -verify-diagnostics %s -canonicalize
+
+llvm.func @malloc(i64) -> !llvm.ptr
+func.func @func2(%arg0: index, %arg1: memref<13x13xi64>, %arg2: index) {
+  %cst_7 = arith.constant dense<1526248407> : vector<1xi64>
+  %1 = llvm.mlir.constant(1 : index) : i64
+  %101 = vector.insert %1, %cst_7 [0] : i64 into vector<1xi64>
+  vector.print %101 : vector<1xi64>
+  return
+}
```
Comment (joker-eph) [inline mlir/test/mlir-opt/issue-74236.mlir:10]: Can we add this to the relevant canonicalize test that already covers this pattern and remove this file please?

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 2]
```diff
@@ -2908,6 +2911,16 @@ class InsertOpConstantFolder final : public OpRewritePattern<InsertOp> {
     rewriter.replaceOpWithNewOp<arith::ConstantOp>(op, newAttr);
     return success();
   }
+
+private:
+  /// Helper function to cast or convert attribute to the expected type.
+  Attribute castOrConvertAttr(Attribute attr, Type expectedType) const {
```
Comment (joker-eph) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2917]: Nit: wouldn't a more accurate name be "convertIntegerAttr"? There is no "cast" here as far as I can tell, and this does nothing for non integer attribute (the doc should be more clear on this)

File: mlir/test/mlir-opt/issue-74236.mlir [hunk 3]
```diff
@@ -0,0 +1,10 @@
+// RUN: mlir-opt -split-input-file -verify-diagnostics %s -canonicalize
+
+llvm.func @malloc(i64) -> !llvm.ptr
+func.func @func2(%arg0: index, %arg1: memref<13x13xi64>, %arg2: index) {
+  %cst_7 = arith.constant dense<1526248407> : vector<1xi64>
+  %1 = llvm.mlir.constant(1 : index) : i64
```
Comment (joker-eph) [inline mlir/test/mlir-opt/issue-74236.mlir:6]: The fact that we allow the `llvm.mlir.constant` to materialize an integer with another type is a bug I believe (unfortunately this creeped up and is quite involved to fix now, I tried... :( )

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 4]
```diff
@@ -2896,10 +2896,13 @@ class InsertOpConstantFolder final : public OpRewritePattern<InsertOp> {
...
-      llvm::append_range(insertedValues, denseSource.getValues<Attribute>());
-    else
-      insertedValues.push_back(sourceCst);
+    Type destEltType = destTy.getElementType();
+    if (auto denseSource = llvm::dyn_cast<DenseElementsAttr>(sourceCst)) {
+      for (auto value : denseSource.getValues<Attribute>())
+        insertedValues.push_back(castOrConvertAttr(value, destEltType));
+    } else {
+      insertedValues.push_back(castOrConvertAttr(sourceCst, destEltType));
+    }
```
Comment (joker-eph) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2905]: Please document the code block as working around the weirdness of LLVM constant which can hold an attribute with a different type than the return type.

**Review by joker-eph — APPROVED** [review]
LG with a few comments

File: mlir/test/mlir-opt/issue-74236.mlir (hunk 1 above)
Comment (Rajveer100) [reply mlir/test/mlir-opt/issue-74236.mlir]: Sure, let me know if the new change works well.

**Review by joker-eph — APPROVED** [review]

**Pull request merged**